dumbo/src/pdu/udp.rs

```diff
@@ -0,0 +1,320 @@
+        if let Some((src_addr, dst_addr)) = verify_checksum {
+            if datagram.compute_checksum(src_addr, dst_addr) != 0 {
+                return Err(Error::Checksum);
+            }
+        }
+
+        Ok(datagram)
+    }
+
+    /// Returns the source port of the UDP datagram.
+    #[inline]
+    pub fn source_port(&self) -> u16 {
+        self.bytes.ntohs_unchecked(SOURCE_PORT_OFFSET)
+    }
+
+    /// Returns the destination port of the UDP datagram.
+    #[inline]
+    pub fn destination_port(&self) -> u16 {
+        self.bytes.ntohs_unchecked(DESTINATION_PORT_OFFSET)
+    }
+
+    /// Returns the length of the datagram from its header.
+    #[inline]
+    pub fn len(&self) -> u16 {
+        self.bytes.ntohs_unchecked(LENGTH_OFFSET)
+    }
+
+    /// Returns the checksum value of the packet.
+    #[inline]
+    pub fn checksum(&self) -> u16 {
+        self.bytes.ntohs_unchecked(CHECKSUM_OFFSET)
+    }
+
+    /// Returns the payload of the UDP datagram as an `[&u8]` slice.
+    #[inline]
+    pub fn payload(&self) -> &[u8] {
+        // Payload offset is header len.
+        self.bytes.split_at(PAYLOAD_OFFSET).1
+    }
+
+    /// Computes the checksum of a UDP datagram.
+    #[inline]
+    pub fn compute_checksum(&self, src_addr: Ipv4Addr, dst_addr: Ipv4Addr) -> u16 {
+        crate::pdu::compute_checksum(&self.bytes, src_addr, dst_addr, ChecksumProto::Udp)
+    }
+}
+
+impl<'a, T: NetworkBytesMut> UdpDatagram<'a, T> {
+    /// Writes an incomplete UDP datagram, which is missing the `checksum`, `src_port` and `dst_port` fields.
+    ///
+    /// # Arguments
+    ///
+    /// * `buf` - A buffer containing `NetworkBytesMut` representing a datagram.
+    /// * `payload` - Datagram payload.
+    #[inline]
+    pub fn write_incomplete_datagram(buf: T, payload: &[u8]) -> Result<Incomplete<Self>, Error> {
+        let mut packet = match UdpDatagram::from_bytes(buf, None) {
+            Ok(packet) => packet,
+            Err(e) => return Err(e),
+        };
+
+        let len = payload.len() + UDP_HEADER_SIZE;
+
+        // TODO working with IPv4 only for now
+        if len > IPV4_MAX_UDP_PACKET_SIZE {
+            return Err(Error::PayloadTooBig);
+        }
+
+        packet.bytes.shrink_unchecked(len as usize);
+        packet.set_payload(payload).set_len(len as u16);
+
+        Ok(Incomplete::new(packet))
+    }
+
+    /// Sets the source port of the UDP datagram.
+    #[inline]
+    pub fn set_source_port(&mut self, src_port: u16) -> &mut Self {
+        self.bytes.htons_unchecked(SOURCE_PORT_OFFSET, src_port);
+        self
+    }
+
+    /// Sets the destination port of the UDP datagram.
+    #[inline]
+    pub fn set_destination_port(&mut self, dst_port: u16) -> &mut Self {
+        self.bytes
+            .htons_unchecked(DESTINATION_PORT_OFFSET, dst_port);
+        self
+    }
+
+    /// Sets the payload of the UDP datagram.
+    #[inline]
+    pub fn set_payload(&mut self, payload: &[u8]) -> &mut Self {
```

Comment:
This method is a bit tricky to use, and it can panic because `copy_from_slice` panics if the slices differ in length. How about replacing this with a `payload_mut(&mut self) -> &mut [u8]` that returns the payload bytes as a mutable slice.

Reply:
Sure, no problems. So, this will leave the responsibility to the caller.
One side effect is that `copy_from_slice` will be moved to `write_incomplete_datagram`